### PR TITLE
bpf: Expand LPM key for future usage.

### PR DESCRIPTION
**Pull request by joestringer**
Expand the lpm keys so that in future we can place port/protocol inside for supporting CIDR-dependent L4, without breaking the map format again.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/3173)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Review by ianvernon — COMMENTED** [review]
How will this affect upgrade?

**Comment by joestringer** [comment]
@ianvernon I'm hoping the upgrade test will tell me ;-)

**Comment by joestringer** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
Blocked by https://github.com/cilium/cilium/issues/3182?

**Comment by joestringer** [comment]
Saw this testing locally:

```
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Command execution failed" cmd="/var/lib/cilium/bpf/join_ep.sh /var/lib/cilium/bpf /var/run/cilium/state /var/run/cilium/state/173_next lxc6f3e6 true 173" containerID=f3ca83c290 endpointID=173 
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Join EP id=/var/run/cilium/state/173_next ifname=lxc6f3e6"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="kernel version:  Linux runtime 4.9.13-040913-generic #201702260631 SMP Sun Feb 26 11:33:27 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="clang version:  clang version 3.8.1 (tags/RELEASE_381/final) Target: x86_64-unknown-linux-gnu Thread model: posix InstalledDir: /usr/local/bin"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="'probe' is not a recognized processor for this target (ignoring processor)"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="'probe' is not a recognized processor for this target (ignoring processor)"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Map specification differs from pinned file!"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg=" - Size key:     24 (obj) != 20 (pin)"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Map 'cilium_cidr_ingress6_173' self-check failed!"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Error loading maps into kernel!"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Error fetching ELF ancillary data!"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="Note: 8 bytes struct bpf_elf_map fixup performed due to size mismatch!"
Mar 16 21:31:23 runtime cilium-agent[2674]: level=warning msg="mv: cannot stat '/sys/fs/bpf/tc/globals/cilium_calls_173.old': Operation not permitted"
```

Looks like a couple of cidr instances were missed. Will update.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Adding padding even to the XDP CIDR map entries, just because the userspace implementation has one cidrmap for all LPM cases so it's easier to reuse the same map implementation and just update the keys to be the same in the datapath.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by ianvernon** [comment]
Needs rebase, but that's just to fix up the BPF SHA.
Can someone from @cilium/bpf please review? This is marked as a 1.0-blocker.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: bpf/lib/maps.h [hunk 1]
```diff
@@ -121,18 +121,22 @@ struct bpf_elf_map __section_maps tunnel_endpoint_map = {
 
 struct bpf_lpm_trie_key6 {
 	struct bpf_lpm_trie_key lpm_key;
+	__u16 pad1;
+	__u16 pad2;
```
Comment (borkmann) [inline bpf/lib/maps.h]: Why not making this __u32?

File: bpf/bpf_xdp.c [hunk 2]
```diff
@@ -107,6 +107,7 @@ static __always_inline int check_v4(struct xdp_md *xdp)
 #ifdef CIDR4_FILTER
 	__builtin_memcpy(pfx.lpm.data, &ipv4_hdr->saddr, sizeof(pfx.addr));
 	pfx.lpm.prefixlen = 32;
+	pfx.pad = 0;
```
Comment (borkmann) [inline bpf/bpf_xdp.c]: I think here may be a bug: in the `lpm_v{4,6}_key` case you also add this in front of the key, therefore prefixlen needs to grow as well by as in the other lpm case due to the extra zeroed prefix.

File: bpf/lib/maps.h [hunk 3]
```diff
@@ -121,18 +121,22 @@ struct bpf_elf_map __section_maps tunnel_endpoint_map = {
...
 	__be32 lpm_addr;
 };
 
 static __always_inline int lpm6_map_lookup(struct bpf_elf_map *map,
 					   union v6addr *addr, __u32 prefix)
 {
-	struct bpf_lpm_trie_key6 key = { { prefix }, *addr };
+	struct bpf_lpm_trie_key6 key = { { prefix + sizeof(__u32) }, 0, 0, *addr };
```
Comment (borkmann) [inline bpf/lib/maps.h]: Shouldn't this be: sizeof(__u32) * 8 ? Call-sites use prefix value of either 32 or 128, and if we extend it by 4 bytes, it should sum up to either 64 or 160 eventually, no?

File: bpf/lib/maps.h (hunk 1 above)
Comment (joestringer) [reply bpf/lib/maps.h]: Originally I was thinking in terms of port + proto, which would need to be split into (at least) two fields anyway. I can fix this up for simplicity.

File: bpf/lib/maps.h (hunk 3 above)
Comment (joestringer) [reply bpf/lib/maps.h]: right, I just thought about this last night and apparently didn't think clearly enough. I'll fix it.

**Comment by joestringer** [comment]
test-me-please

**Comment by borkmann** [comment]
Strange, error seems to be:
```
18:34:16 # github.com/cilium/cilium/daemon
18:34:16 /tmp/go-build751993867/github.com/cilium/cilium/daemon/_test/_obj_test/bindata.go:1872:2: duplicate key "bpf/bpf_xdp.c" in map literal
18:34:16 /tmp/go-build751993867/github.com/cilium/cilium/daemon/_test/_obj_test/bindata.go:1873:2: duplicate key "bpf/bpf_xdp.c" in map literal
18:34:16 FAIL	github.com/cilium/cilium/daemon [build failed]
18:34:16 Makefile:40: recipe for target 'tests-ginkgo-real' failed
18:34:16 make: *** [tests-ginkgo-real] Error 1
18:34:16 Makefile:26: recipe for target 'tests-ginkgo' failed
18:34:16 make: *** [tests-ginkgo] Error 2
```

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]
LGTM

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
I'm pursuing a different approach to adding support for CIDR maps (outlined in #1684), so this PR will be unnecessary.

**Pull request closed without merging**